multicluster/controllers/multicluster/label_identity_controller.go

```diff
@@ -0,0 +1,347 @@
+	r.labelMutex.Lock()
+	defer r.labelMutex.Unlock()
+
+	if originalLabel, isCached := r.podLabelCache[podNamespacedName]; isCached {
+		if podNames, ok := r.labelToPodsCache[originalLabel]; ok && len(podNames) == 1 && podNames.Has(podNamespacedName) {
+			klog.V(2).InfoS("Label no longer exists in the cluster", "label", originalLabel)
+			// No other workers can potentially be creating a ResourceExport for this label, because
+			// before this current deletion, at least one Pod is referring to the label; it is therefore
+			// not a new label. Nor can the label's ResourceExport under deletion since only the last Pod
+			// referring to the label can trigger deletion.
+			return originalLabel
+		} else if ok {
+			// No ResourceExport deletion is needed. Update the caches immediately.
+			delete(r.podLabelCache, podNamespacedName)
+			podNames.Delete(podNamespacedName)
+		}
+	}
+	return ""
+}
+
+// updateLabelCache updates podLabelCache and labelToPodsCache once the ResourceExport
+// create/update operation is successful.
+func (r *LabelIdentityReconciler) updateLabelCache(podNamespacedName, updatedLabel string) {
+	r.labelMutex.Lock()
+	defer r.labelMutex.Unlock()
+
+	r.podLabelCache[podNamespacedName] = updatedLabel
+	if _, ok := r.labelToPodsCache[updatedLabel]; !ok {
+		r.labelToPodsCache[updatedLabel] = sets.NewString(podNamespacedName)
+	}
+	r.labelToPodsCache[updatedLabel].Insert(podNamespacedName)
+}
+
+// deleteFromLabelCache deletes a label from the podLabelCache and labelToPodsCache once
+// the ResourceExport delete operation is successful.
+func (r *LabelIdentityReconciler) deleteFromLabelCache(podNamespacedName, deletedLabel string) {
+	r.labelMutex.Lock()
+	defer r.labelMutex.Unlock()
+
+	delete(r.podLabelCache, podNamespacedName)
+	if podNames, ok := r.labelToPodsCache[deletedLabel]; ok {
+		podNames.Delete(podNamespacedName)
+		delete(r.labelToPodsCache, deletedLabel)
+	}
+}
+
+func (r *LabelIdentityReconciler) createLabelIdentityResExport(ctx context.Context, normalizedLabel string) error {
+	setLabelResExportCreating := func() {
+		r.resourceExportMutex.Lock()
+		defer r.resourceExportMutex.Unlock()
+		r.labelAddsInProgress.Insert(normalizedLabel)
+	}
+	setLabelResExportCreating()
+	resExportName := getResourceExportNameForLabelIdentity(r.localClusterID, normalizedLabel)
+	labelResExport := r.getLabelIdentityResourceExport(r.remoteCommonArea.GetNamespace(), resExportName, normalizedLabel)
+	err := r.remoteCommonArea.Create(ctx, labelResExport, &client.CreateOptions{})
+
+	r.resourceExportMutex.Lock()
+	defer r.resourceExportMutex.Unlock()
+
+	r.labelAddsInProgress.Delete(normalizedLabel)
+	if err != nil && !apierrors.IsAlreadyExists(err) {
+		return err
+	}
+	return nil
+}
+
+func (r *LabelIdentityReconciler) deleteLabelIdentityResExport(ctx context.Context, normalizedLabel string) error {
+	setLabelResExportDeleting := func() {
+		r.resourceExportMutex.Lock()
```

Comment:
Will it be simpler to use sync.Map?

Reply:
`sync.Map` is actually the first thing that came to my mind. But according to the documentation, it recommends for use case like this one we simply put a mutex around a map to be more efficient in terms of no type casting. `sync.Map` is only optimized for one write and multiple reads.